From patchwork Thu Aug 30 23:05:37 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10582973
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CCF3E5A4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 30 Aug 2018 23:06:22 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id AECC52B821
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 30 Aug 2018 23:06:22 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id A0B142C217; Thu, 30 Aug 2018 23:06:22 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-10.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,USER_IN_DEF_DKIM_WL autolearn=ham
	version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id B66902B821
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 30 Aug 2018 23:06:08 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727086AbeHaDKh (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 30 Aug 2018 23:10:37 -0400
Received: from mail-ua1-f74.google.com ([209.85.222.74]:56459 "EHLO
        mail-ua1-f74.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1725836AbeHaDKh (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 30 Aug 2018 23:10:37 -0400
Received: by mail-ua1-f74.google.com with SMTP id i39-v6so4474236uaa.23
        for <kvm@vger.kernel.org>; Thu, 30 Aug 2018 16:06:06 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=VBF/PLz97tP9iYKOsI6/naNNY6BTW/T6YWnMieBfLXw=;
        b=kbd0yS39N5FcBC4J8ufyhAPcnLe43c+2HS63+s0MDs4k5eKbyneLRD8oKsCiiWjWb1
         TDVDX2X5u63nQcDtDjnHr1+umvLiZbelNgII6Vf/Oy2GPIjNKwshjYOFMMHRqngSEs+0
         3tZn9f0T5Cn7VE99GCHid1u8iTcNA/Xmm9kZRUFsOy8PYK1+FkE/x1FQvWtNq7foevMw
         cZW1x0sY7gpsT3vKYDFLwtmz4RWwkKBhEb79A0rAKmzG+nxZfEoEiYF479oJUtolOOIf
         QDWYaHKxe+oyVP+kLAepMFCS9kZbv+pklCQjvWaxXBcf0RRKhdBDs0uVx/DIlhwrs+0+
         WKHw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=VBF/PLz97tP9iYKOsI6/naNNY6BTW/T6YWnMieBfLXw=;
        b=rmvX0XQAP7auJ0HNDBbdZkk6hOjhu9JkTMY8ivGOv0l59lFmBjIoqqK62Nup9AAQCP
         vr53V8mGBVCgiX8sGin/h+U6T2qN6LLgW3twFuRPEb6CbRsWs7HkPdqQmfTWzWB3msoL
         aFE/xv01GKxZv4aM+JRFy7FRPWIrgUkAd8qgV0EouCtYlSMNoBX1TOVNvmEh9OYc6NdA
         wttNz/HNpA4OUB008vvgFh6jecy01bj4NptW5Pi49HKT2M/oN6D0OTdnIXX2BXW3ARoK
         7x97wVaDjRJssQRktyU/NUa/nVV6B2eLfGkuow4fmqpGAfAsf0nQs7YatLtIdWY5dVs6
         hZ7A==
X-Gm-Message-State: APzg51C37IbzYpnFZYaaDRUfTINcio/aQO0Cf6RsEHpT/EW/14AJRjQY
        H/U7EnAo19QbNAbgryJAsUQ938e0KcXFWXk4YIkoTgNMophhVFv1V5qwyio5fl7X/Ooe17NQZCe
        LsOuqil83NzWA73ik73hJIbBXBu2fEZ0SnsfIamddDT78YWO9DD+fAVzpF03XL0Q=
X-Google-Smtp-Source: 
 ANB0Vdb82fwPJUF2RammrAs5sLEOR+c6waZfy/LKcdpwIHKivzNZlH+oWsa/foxwvYV9raTuOwxmxS1MmqNCaQ==
X-Received: by 2002:ab0:12e8:: with SMTP id
 o40-v6mr6006120uac.50.1535670365764;
 Thu, 30 Aug 2018 16:06:05 -0700 (PDT)
Date: Thu, 30 Aug 2018 16:05:37 -0700
Message-Id: <20180830230537.80356-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.19.0.rc0.228.g281dcd1b4d0-goog
Subject: [PATCH] KVM: nVMX: Unrestricted guest mode requires EPT
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Andrew Thornton <andrewth@google.com>,
        Peter Shier <pshier@google.com>,
        Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

As specified in Intel's SDM, do not allow the L1 hypervisor to launch
an L2 guest with the VM-execution controls for "unrestricted guest" or
"mode-based execute control for EPT" set and the VM-execution control
for "enable EPT" clear.

Note that the VM-execution control for "mode-based execute control for
EPT" is not yet virtualized by kvm.

Reported-by: Andrew Thornton <andrewth@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 arch/x86/include/asm/vmx.h |  1 +
 arch/x86/kvm/vmx.c         | 13 +++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/arch/x86/include/asm/vmx.h b/arch/x86/include/asm/vmx.h
index 9527ba5d62da..665632a4b54b 100644
--- a/arch/x86/include/asm/vmx.h
+++ b/arch/x86/include/asm/vmx.h
@@ -78,6 +78,7 @@
 #define SECONDARY_EXEC_RDSEED_EXITING		0x00010000
 #define SECONDARY_EXEC_ENABLE_PML               0x00020000
 #define SECONDARY_EXEC_XSAVES			0x00100000
+#define SECONDARY_EXEC_MODE_BASED_EPT_EXEC	0x00400000
 #define SECONDARY_EXEC_TSC_SCALING              0x02000000
 
 #define PIN_BASED_EXT_INTR_MASK                 0x00000001
diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 1d26f3c4985b..2bf990b5848f 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -11719,6 +11719,16 @@ static int nested_vmx_check_pml_controls(struct kvm_vcpu *vcpu,
 	return 0;
 }
 
+static int nested_vmx_check_ept_enable(struct kvm_vcpu *vcpu,
+				       struct vmcs12 *vmcs12)
+{
+	if ((nested_cpu_has2(vmcs12, SECONDARY_EXEC_UNRESTRICTED_GUEST) ||
+	     nested_cpu_has2(vmcs12, SECONDARY_EXEC_MODE_BASED_EPT_EXEC)) &&
+	    !nested_cpu_has_ept(vmcs12))
+		return -EINVAL;
+	return 0;
+}
+
 static int nested_vmx_check_shadow_vmcs_controls(struct kvm_vcpu *vcpu,
 						 struct vmcs12 *vmcs12)
 {
@@ -12339,6 +12349,9 @@ static int check_vmentry_prereqs(struct kvm_vcpu *vcpu, struct vmcs12 *vmcs12)
 	if (nested_vmx_check_pml_controls(vcpu, vmcs12))
 		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
 
+	if (nested_vmx_check_ept_enable(vcpu, vmcs12))
+		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
+
 	if (nested_vmx_check_shadow_vmcs_controls(vcpu, vmcs12))
 		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
 
